gossip/tips_cache.c

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2018 IOTA Stiftung
+ * https://github.com/iotaledger/entangled
+ *
+ * Refer to the LICENSE file for licensing information
+ */
+
+#include "gossip/tips_cache.h"
+#include "utils/handles/rand.h"
+
+/*
+ * Private functions
+ */
+
+static retcode_t tips_cache_fifo_add(hash243_set_t* const set,
+                                     size_t const capacity,
+                                     flex_trit_t const* const tip) {
+  retcode_t ret = RC_OK;
+
+  if (set == NULL || tip == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  if (hash243_set_size(set) >= capacity) {
+    if ((ret = hash243_set_remove_entry(set, *set)) != RC_OK) {
+      return ret;
+    }
+  }
+
+  return hash243_set_add(set, tip);
+}
+
+static retcode_t tips_cache_random_tip_from_set(hash243_set_t* const set,
+                                                flex_trit_t* const tip) {
+  hash243_set_entry_t* iter = NULL;
+  hash243_set_entry_t* tmp = NULL;
+
+  if (set == NULL || tip == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  int index = rand_handle_rand_interval(0, hash243_set_size(set));
+  HASH_ITER(hh, *set, iter, tmp) {
+    if (index-- == 0) {
+      memcpy(tip, iter->hash, FLEX_TRIT_SIZE_243);
+      break;
+    }
+  }
+
+  return RC_OK;
+}
+
+/*
+ * Public functions
+ */
+
+retcode_t tips_cache_init(tips_cache_t* const cache, size_t const capacity) {
+  if (cache == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  cache->tips = NULL;
+  cache->solid_tips = NULL;
+  cache->capacity = capacity;
+  rw_lock_handle_init(&cache->lock);
```

Comment:
Might be worth adding a note here that we could split this into two distinct locks which might net us a performance improvement

Reply:
I'll add the second lock as part of this PR